docs/madr/decisions/056-policy-matching-with-real-resources.md

```diff
@@ -0,0 +1,185 @@
+
+The approach has a number of disadvantages:
+
+* even though it works well for "from" policies, using it for "to" policies doesn't feel natural
+* when referencing the real MeshHTTPRoute we had to use subset with artificial `__rule-matches-hash__` key
+* adding support for `MeshService` forces us to introduce another artificial key like `__resource-name__`
+* Inspect API with subsets is tricky when referencing real resources (like MeshHTTPRoute or MeshService)
+
+We need a new format that'd make things easier when it comes to referencing real resources in `to[]`.
+
+## Considered Options and Decision Outcome
+
+No changes for "from" policies, both policy matching algorithm and Inspect API stays the same.
+
+For "to" policies the algorithm is going to return the additional structure:
+
+```go
+type ToRules struct {
+    Rules         Rules 
+    ResourceRules map[UniqueResourceKey]ResourceRule // new field
+}
+
+type ResourceRule struct {
+    Resource core_model.ResourceMeta
+    Conf     interface{}
+    Origin   []core_model.ResourceMeta
+}
+```
+
+When targetRef has kind `MeshService` and uses `labels` we're going to produce multiple `ResourceRule` entities
+for each real `MeshService` that satisfies `labels`, i.e for the given `targetRef`:
+
+```yaml
+to:
+  - targetRef:
+      kind: MeshService
+      labels:
+        k8s.kuma.io/namespace: finance
+    conf: conf1
+```
+
+we are going to fetch all `MeshService` resources, filter them by `k8s.kuma.io/namespace: finance` and produce N rules:
+
+```yaml
+resourceRules:
+  - resource:
+      name: finance-backend.finance
+      mesh: mesh-1
+    conf: conf-1
+    origin: [...]
+  - resource:
+      name: finance-frontend.finance
+      mesh: mesh-1
+    conf: conf-1
+    origin: [...]
+  - resource:
+      name: finance-db.finance
+      mesh: mesh-1
+    conf: conf-1
+    origin: [...]
+```
+
+Policy plugins should implement a fallback logic when trying to get a conf for the given `MeshService{name: backend}`:
+
+1. Lookup `MeshService{name: backend}` in the new `ToRules.ResourceRules`
+2. If it exists then the real MeshService `backend` exists in the cluster and we can use `ResourceRule.Conf`
+3. If it's absent then fallback to legacy `ToRules.Rules` and get conf for `kuma.io/service: backend` tag
+
+Eventually we're going to delete `ToRules.Rules`.
+
+When it comes to the Inspect API, we're going to add a new field `to `InspectRule`
+
+```yaml
+InspectRule:
+  type: object
+  required: [type]
+  properties:
+    type:
+      type: string
+      example: MeshRetry
+      description: the type of the policy
+    proxyRule:
+      description: a rule that affects the entire proxy
+      $ref: '#/components/schemas/ProxyRule'
+    toRules: # should be removed eventually
+      type: array
+      description: a set of rules for the outbounds of this proxy
+      items:
+        $ref: '#/components/schemas/Rule'
+    resourceRules: # new field
```

Comment:
shouldn't this be `toResourceRules`? Do you foresee this being used for anything other than `to` section?